Do not use non-wf input expectations from fudge when checking function calls

ShoyuVanilla opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/150316)*
<!-- homu-ignore:end -->

cc https://github.com/rust-lang/rust/issues/149379

r? lcnr 

# FCP: Do not use non-wf input expectations from fudge when checking function calls

## What is fudging?

https://github.com/rust-lang/rust/blob/71e00273c0921e1bc850ae8cc4161fbb44cfa848/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L168-L170

Consider this coercion site:

`let _: Box<dyn Fn(&str) -> usize> = Box::new(|s| s.len());`

We rely on the expectation to eagerly infer the type of `s` to be `&str`. However, `dyn Fn(&str) -> usize` is not a valid type as the argument of `Box::new`, as it is not `Sized`. 

*Fudging* is the mechanism we use to propagate the expectation through the `Box::new` call without constraining its generic parameter.

Fudging computes the expected argument types by acting as if we're able to propagate the expected return type directly through the function, without any coercions on the return site.

Given that we may actually want to coerce afterwards, we cannot actually commit any constraints here. We therefore compute the expectations for the function arguments in a `probe` and rely on *fudging* to be able to name any inference variables created inside of the probe.

After the fudging step, we weaken the resulting expectation if it is an unsized type in the following lines:
https://github.com/rust-lang/rust/blob/71e00273c0921e1bc850ae8cc4161fbb44cfa848/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L354
https://github.com/rust-lang/rust/blob/71e00273c0921e1bc850ae8cc4161fbb44cfa848/compiler/rustc_hir_typeck/src/expectation.rs#L77-L89

Because function arguments must be `Sized`, this weakening prevents us from applying wrong, unsized coercions to them.

## How fudging currently goes wrong

We have an opened issue for tracking such cases: https://github.com/rust-lang/rust/issues/149379.

Broadly, the failures seem to fall into two buckets.

### We may end up with non–well-formed expectations

Fudging can produce an expected type that is not well-formed.

That would eventually result in an error failing the well-formedness check, either when we do the coercion with the expected argument types, or when we select the remaining obligations.

```rust
fn foo<T>(x: (T, ())) -> Box<T> {
    Box::new(x.0)
}

fn main() {
    // We use `(dyn Send, ())` as the expectation the argument.
    let _: Box<dyn Send> = foo(((), ()));
}
```

### Weakening fudged expectation is not covering all the cases 

```rust
fn field_to_box<T>(x: &(T,)) -> &T {
    &x.0
}
fn main() {
    // `Expectation::rvalue_hint` only checks whether the whole argument
    // itself is `Sized`. It does not check whether the function requires
    // its generic parameters to be `Sized`.
    let _: &dyn Send = field_to_box(&(1,));
}
```

## What this PR fixes

### One of the problematic cases of the issue

This PR fixes the first case, by simply checking well-formedness of the each expected argument types inside the fudge scope.

This is a reasonable change because:
- Non well-formed expectation would result in a well-formedness error so not using such expectation wouldn't make any previously compiled code being not compiled anymore
- Dropping a non well-formed expectation does not mean we stop providing expectations for argument coercions altogether.
  If fudging fails, we still fall back to using the types from the function signature as expectations in the usual path:
  https://github.com/rust-lang/rust/blob/71e00273c0921e1bc850ae8cc4161fbb44cfa848/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs#L330-L336

#### Related tests

- Fixes [tests/ui/coercion/fudge-inference/fn-ret-trait-object-propagated-to-inputs-issue-149379-1.rs](https://github.com/rust-lang/rust/pull/150316/commits/5668ad597d8293dcfd8917ca6e8f78d2c06555d3#diff-1468a6d8495f7adfb4a64508f002bb934c13d13871662de6efd60433649401fd)

### Limited for `-Znext-solver`

Separately (and not directly tied to the above issue), this PR also fixes a next-solver regression tracked at: https://github.com/rust-lang/trait-system-refactor-initiative/issues/259.

That regression was introduced by https://github.com/rust-lang/rust/pull/149320, which started normalizing expected function input types inside the fudge scope.
Because all inference-variable relationships and pending obligations are dropped out when we exit the fudge scope, we drop the ambiguous goal used to normalize, leaving us with an unconstrained inference variable in the expectation.

This PR fixes that by normalizing the expectation outside the fudge scope, so the resulting normalization obligations are preserved to the `InferCtxt`'s `ObligationCtxt`.

#### Related tests

- Fixes [tests/ui/traits/next-solver/fudge-inference/do-not-drop-ambig-normalization.rs](https://github.com/rust-lang/rust/pull/150316/files#diff-42e97f178fbdee7c3405ae12409eb0bca4eec92488971c703b26c083eadf728a)

## Does this PR break anything

I don't think there should be any breakage affecting the old solver.

The separate expectation normalization change only affecting the new solver does break an existing [test](https://github.com/rust-lang/rust/pull/150316/files#diff-3b946a09e063aad2a4fa6b0893508d5ffab78763b8465abfe1f689d349fda815).
This is unfortunate but I think this change should be done because

- The broken test also doesn't compile with the old solver
- The expectation normalization change is necessary to compile stuff supported on stable

## What this PR doesn't fix

This PR doesn't fix the second case -- *Weakening fudged expectation is not covering all the cases*.

@lcnr has [suggested](https://rust-lang.zulipchat.com/#narrow/channel/144729-t-types/topic/expectations.20from.20fudging.20are.20a.20mess.20.23149379/near/560625205) the following solution for that problem: 
> check whether a function where-bound errors without an out coercion, if so, weaken the expectation to `ExpectRvalueLikeUnsized`

I experimented with this and it works well in many cases.
However, on the old solver, checking where-bounds cannot reliably be treated as speculative: if we hit an overflow while checking the where-bounds, the old solver can fail the entire compilation rather than merely treating the check as failed and relaxing the expectation.

Since this second class of issues affects both the old solver and the next-solver, it seems preferable to keep the conservative behavior for now, at least until the next-solver is stabilized, rather than introducing a next-solver-only relaxation that might create new regressions and complicate stabilization efforts.

#### Related tests

- Does NOT Fix [tests/ui/coercion/fudge-inference/expectated-input-not-satisfying-fn-bounds-issue-89299.rs](https://github.com/rust-lang/rust/pull/150316/files#diff-fdcfa8ab660c052dbe246db279d167ea8a309bfe10ca6163f7fa1836be2b30d6)
- Does NOT Fix [tests/ui/coercion/fudge-inference/expectated-input-not-satisfying-fn-bounds-issue-149881.rs](https://github.com/rust-lang/rust/pull/150316/files#diff-1ccbb181cbf164841ca5af350ecf903c802a4854bda309e83e91c3b917809a55)
- Does NOT Fix [tests/ui/coercion/fudge-inference/fn-ret-trait-object-propagated-to-inputs-issue-149379-3.rs](https://github.com/rust-lang/rust/pull/150316/files#diff-b12e01cc3c265db42f135d67425d8b2bd0d9c44c680b3e8c49d1f845a0b25d09)

lcnr committed:
fix fudging

lcnr commented on compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:
```diff
@@ -290,6 +292,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                            self.param_env,
+                            ty::ClauseKind::WellFormed(ty.into()),
+                        ));
+                    }
```
why this?

lcnr reviewed (commented):
I feel like this PR is currently doing multiple somewhat distinct changes in the same commit. Please split them up into separate commits (or PRs even)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150316/99ff3fbb86658b427f5dd7daaae8db5626a63c26..bcc851dcee86bac7df3165b2165032680fba70f2)*

ShoyuVanilla committed:
Add tests for issue number 89299 and 149881

ShoyuVanilla committed:
Do not use ill-formed input expectations when checking function calls

lcnr reviewed (commented):
the second change isn't necessary to avoid any breakage when switching solvers, is it?

I think to fix https://github.com/rust-lang/trait-system-refactor-initiative/issues/259 we should just do https://github.com/rust-lang/rust/commit/f5acf220a90c32d4be4441b07340addb2a245266

I think these sorts of improvements are generally good, but don't like improvements which are only limited to the new solver. I would drop the second commit and then get back to that one once the new solver is stable

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150316/a6acf0f07f0ed1c12e26dc0db3b9bf1d0504a0bb..effd66a1b7b339aff5f4a377770e768c92799b16)*

ShoyuVanilla commented:
> Yeah, moving generalization outside of the fudge fixes the issue but it regresses back an improvement made by next-solver as said in [#t-types/trait-system-refactor > diesel benchmark doesn't compile @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/364551-t-types.2Ftrait-system-refactor/topic/diesel.20benchmark.20doesn't.20compile/near/564190256)
> 
> I'll try combining it with the third commit of this PR and if it doesn't fix the regression, I think this should be closed and revisited once the new solver is stabilized 😄

Oops, the third commit + https://github.com/rust-lang/rust/commit/f5acf220a90c32d4be4441b07340addb2a245266 still regresses back `tests/ui/traits/next-solver/lazy-nested-obligations-2.rs` 😅 Closing this for now

lcnr commented:
The first change is also old solver, is it not? That one seems cool and I think I'd like to FCP it rn

ShoyuVanilla commented on compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs:
```diff
@@ -305,7 +305,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    ocx.try_evaluate_obligations().is_empty()
+                });
+                well_formed.then_some(expected_input_tys)
+            });
```
Yeah, this is very absurd 😅 I guess I was trying to do more things but just ended up something equivalent to filter

ShoyuVanilla commented:
I've dropped the last commit and fixed the filter thing.

Unfortunately, the wf commit + https://github.com/rust-lang/rust/commit/f5acf220a90c32d4be4441b07340addb2a245266 still regresses back `tests/ui/traits/next-solver/lazy-nested-obligations-2.rs` 🤔 

@rustbot ready

lcnr commented on tests/ui/traits/next-solver/lazy-nested-obligations-2.rs:
```diff
@@ -1,5 +1,8 @@
 //@ compile-flags: -Znext-solver
-//@ check-pass
+
+// FIXME(#149379): This fails on the old solver. The next solver used to accept it
```
*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150316/2bd7a97871a74d4333bd3edb6564136167ac604b..94e58c09796d793427079f639a5a576e47304479)*

can you use explicit revisions instead of stating "This fails on the old solver"

lcnr reviewed (commented):
r=me after nit

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150316/2bd7a97871a74d4333bd3edb6564136167ac604b..94e58c09796d793427079f639a5a576e47304479)*

ShoyuVanilla committed:
Clarify old solver behavior with revisions in a test

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
This pull request is already queued and waiting for a try build to finish.

<!-- rust-timer: temporary -->

rust-timer commented:
This pull request is already queued and waiting for a try build to finish.

<!-- rust-timer: temporary -->

panstromek commented:
perf triage:

Let's try this again

@rust-timer build b520876afc597b003c72cb107b37378c4c9102cf

rust-timer commented:
Queued b520876afc597b003c72cb107b37378c4c9102cf with parent d0442e2800d356ae282ddcdbe0eff8798fe648b6, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d0442e2800d356ae282ddcdbe0eff8798fe648b6&end=b520876afc597b003c72cb107b37378c4c9102cf).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (b520876afc597b003c72cb107b37378c4c9102cf): [comparison URL](https://perf.rust-lang.org/compare.html?start=d0442e2800d356ae282ddcdbe0eff8798fe648b6&end=b520876afc597b003c72cb107b37378c4c9102cf&stat=instructions:u).

### Overall result: ❌ regressions - please read:


Benchmarking means the PR may be perf-sensitive. It's automatically marked not fit for rolling up. Overriding is possible but disadvised: it risks changing compiler perf.

**Next, please**: If you can, justify the regressions found in this try perf run in writing along with `@rustbot label: +perf-regression-triaged`. If not, fix the regressions and do another perf run. Neutral or positive results will clear the label automatically.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=d0442e2800d356ae282ddcdbe0eff8798fe648b6&end=b520876afc597b003c72cb107b37378c4c9102cf&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.2%, 0.8%]   | 37    |
| Regressions ❌ <br /> (secondary)  | 0.6%  | [0.1%, 2.5%]   | 45    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | 0.5%  | [0.2%, 0.8%]   | 37    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d0442e2800d356ae282ddcdbe0eff8798fe648b6&end=b520876afc597b003c72cb107b37378c4c9102cf&stat=max-rss)
<details>
<summary>Results (secondary -2.6%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-2.6%, -2.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d0442e2800d356ae282ddcdbe0eff8798fe648b6&end=b520876afc597b003c72cb107b37378c4c9102cf&stat=cycles:u)
<details>
<summary>Results (primary 2.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.2% | [2.2%, 2.2%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.2% | [2.2%, 2.2%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d0442e2800d356ae282ddcdbe0eff8798fe648b6&end=b520876afc597b003c72cb107b37378c4c9102cf&stat=size:linked_artifact)
This perf run didn't have relevant results for this metric.

**Bootstrap**: 489.468s -> 488.816s (-0.13%)
**Artifact size**: 395.51 MiB -> 395.62 MiB (0.03%)

panstromek commented:
This is almost exact match, great.

@ShoyuVanilla This PR has caused a perf regression in https://github.com/rust-lang/rust/pull/155024. Is this expected (I assume not, because there's no pre-merge perf run)? Can we somehow justify it? Should we revert?

ShoyuVanilla commented:
I guess this is somewhat expected and could be justified since this fixes some type inference errors and we are doing extra wf checks for it. But @lcnr might think otherwise 🤔

lcnr commented:
I think we should accept this regression. If we already had these these WF checks from day one I would not have removed them for the performance improvement. While it is unfortunate to cause a perf regression for something that fixes a rather rare edge-case, I do think avoiding these rough corners is generally worth it

panstromek commented:
Thanks. Let's mark as triaged then.

@rustbot label: +perf-regression-triaged
